File: app/sources/manual.py
from __future__ import annotations

from typing import Optional
from app.models import SourceDefinition, SourceDocument, SyncProgress, ConfigKeys
from app.sources.base import Source, ProgressCallback


class ManualDocumentSource(Source):
    def __init__(self, source: SourceDefinition) -> None:
        self._source = source

    async def fetch_documents(
        self, progress_cb: Optional[ProgressCallback] = None
    ) -> list[SourceDocument]:
        content = self._source.get_config(ConfigKeys.CONTENT)
        title = self._source.get_config(ConfigKeys.TITLE) or self._source.name

        if progress_cb:
            progress_cb(SyncProgress(phase="fetching", message=f"Loading {title}"))

        return [SourceDocument(
            id=self._source.id,
            text=content,
            tags={"source_name": self._source.name},
            properties={"title": title},
        )]

File: internal/sources/manual.go
// Manual document source — Go port of app/sources/manual.py.
package sources

import (
	"context"

	"github.com/MichalOndrejka/conduit/internal/models"
)

type ManualSource struct {
	src *models.SourceDefinition
}

func (m *ManualSource) FetchDocuments(_ context.Context, progress ProgressCallback) ([]models.SourceDocument, error) {
	content := m.src.GetConfig("Content")
	if content == models.DocumentPlaceholder {
		return nil, nil
	}
	title := m.src.GetConfig("Title")
	if title == "" {
		title = m.src.Name
	}
	if progress != nil {
		progress(models.SyncProgress{Phase: "fetching", Message: "Loading " + title})
	}
	return []models.SourceDocument{{
		ID:   m.src.ID,
		Text: content,
		Tags: map[string]string{
			"source_id":   m.src.ID,
			"source_name": m.src.Name,
		},
		Properties: map[string]string{"title": title},
	}}, nil
}
